### ISSUE TITLE
Non-detection of HIP

### ISSUE THREAD

**Issue by cmelone** [StanfordLegion/gasnet#23]
Hello,

I am building Legion on Tioga@LLNL (has AMD GPUs) with the following configuration:

```
#!/bin/bash
ml PrgEnv-gnu cmake/3.24.2 python/3.10.8 cray-pmi/6.1.3 rocm/4.5.0

export CC=cc
export CXX=CC
export HOST_CC=gcc
export HOST_CXX=g++
export CONDUIT=ofi-slingshot11
export REALM_NETWORKS=gasnetex
export LEGION_DIR=$PWD/legion
export HDF_ROOT="$LEGION_DIR"/language/hdf/install
export USE_HIP=1
export USE_OPENMP=1
export USE_GASNET=1
export USE_HDF=1
export HIP_ARCH=gfx90a
export HIP_PATH=/opt/rocm-4.5.0/hip
export THRUST_PATH=$PWD/thrust
export MAX_DIM=3
```

Everything builds successfully, but I noticed this line while scrolling through the log:

```
  GPUs with AMD HIP API (hip)                        OFF    (not enabled)
```

How can I set my environment to for gasnet to detect AMD GPUs? I am mostly curious if there are any negative implications while running.

**Comment by elliottslaughter** [comment]
This is a question for @PHHargrove @bonachea because I would have expected GASNet to auto-detect HIP given the `module load rocm/4.5.0`. But I'm not sure why it would be going wrong.

**Comment by PHHargrove** [comment]
Here is what I can think of based on the information provided:

My best guess is that the GASNet-EX build recipe in use did not include either of `--enable-memory-kinds` or `--enable-kind-hip`.  Since GPU memory kinds are an "opt in" feature in GASNet-EX, if neither of those is provided then the reported "not enabled" is the _expected_ outcome.

For instance, I see that the `config.ofi-slingshot11.release` provided in the `StanfordLegion/gasnet` repo does not include `--enable-memory-kinds` (to enable auto-detection).  Understandably for a generic configuration, it lacks the options `--enable-kind-hip` and `--enable-kind-cuda-uva` which would cause `configure` to fail if the corresponding dependencies cannot be found.  

I suggest arranging to configure GASNet-EX with `--enable-kind-hip`, and if that is not sufficient then one might also need `--with-hip-home=/opt/rocm-4.5.0/hip`.  I will defer to @elliottslaughter to provide the right mechanism to do that within the scope of Legion's CMake infrastructure.

If `--enable-kind-hip --with-hip-home=/opt/rocm-4.5.0/hip` proves insufficient or fails in a new way, then please attach the entire output from the GASNet-EX `configure` script.

**Comment by elliottslaughter** [comment]
@PHHargrove would there be any complications if we added `--enable-memory-kinds` to all (relevant) conduit configurations? It only enables auto-detect? (I.e., won't break our build if CUDA/HIP aren't available.) Can anything go wrong if it gets turned on when we don't want it?

I realize that's not what we've done historically and maybe we need to set up configurations `slingshot11-cuda` and `slingshot11-hip` to explicitly enable each respective GPU backend. I'll need to consult the Legion team to be sure. Meanwhile, if we can think through the potential implications, that would help us decide.

**Comment by PHHargrove** [comment]
@elliottslaughter My concern with `--enable-memory-kinds` in all relevant configurations would be _incorrect_ detection due to our `configure` script's limited ability to probe "everything".  In particular we've encountered plenty of InfiniBand clusters with CUDA or HIP devices which lack the necessary kernel drivers (e.g. for GDR / ROCmRDMA support) or had other configuration issues which prevented the memory kinds support from working.  This leads to fatal errors at runtime which one must return to the `configure` step to resolve (by disabling the support in GASNet-EX).  This is one of the main reasons why the feature is "opt-in" currently.

However, the case of the Slighshot-11 should not have the sorts of issues I describe above.  So, `slingshot11-cuda` and `slingshot11-hip` do sound like a reasonable approach to me, using `--enable-kind-cuda-uva` and `--enable-kind-hip` respectively.

**Comment by bonachea** [comment]
To expand on what @PHHargrove said, you needn't necessarily create -hip and -cuda variants of [config.ofi-slingshot11.release](https://github.com/StanfordLegion/gasnet/blob/master/configs/config.ofi-slingshot11.release).

IIUC Legion always wants any available GPU offload support, so it should alternatively be safe to simply `--enable-memory-kinds` in `config-ofi-slingshot*` to enable GPU offload auto-detection on HPE Cray EX systems where I _think_ the appropriate drivers can safely be presumed to always be present. The only counter-example I can think of is Cray EX with Intel GPU systems like Aurora, where the drivers might not yet be installed, but the current release of GASNet lacks Intel GPU offload anyhow.

**Comment by elliottslaughter** [comment]
What about Perlmutter CPU nodes? If we build on the head node, then the build might auto-detect CUDA, but then the CUDA driver might not be available on the CPU nodes.

Am I correct in understanding that if the build detects a CUDA/HIP compiler available, but the compute nodes do not have drivers installed, that will cause a crash?

P.S. Whatever the particular answer for Perlmutter, a similar machine with CPU/GPU partitions might be configured differently, so assuming my understanding in the second paragraph is correct, it seems safer to explicitly configure.

**Comment by PHHargrove** [comment]
Perlmutter's CPU nodes are an interesting case that I don't think was sufficiently explained in my comments or Dan's.

When I said there was a fatal run time error, I was not sufficiently clear on the conditions.  In fact, that only occurs when the client (`librealm` in the case of interest) attempts to use GPU memory as a GASNet segment.  Under the realistic assumption that a well-written client would not do that when it has not detected any GPUs, there will be no error.  IOW: I am fairly confident (though I confess I've not tested thist) that a single build of `libgasnet` configured using `--enable-kind-cuda-uva` of `--enable-memory-kinds` will function correctly on **both** partitions of Perlmutter while fully utilizing the memory kinds support on the GPU partition.

The same goes for "if the build detects a CUDA/HIP compiler available, but the compute nodes do not have drivers installed" **provided** "drivers" refers to the `cuda` or `rocm` drivers.  The problematic case is the presence of those GPU device drivers (sufficient to compute) and at least one device they recognize, coupled with lack of the PCI peer-to-peer driver (e.g `nv_peer_mem` or `amdfkd`) needed for GPU-NIC transfers.

**Comment by elliottslaughter** [comment]
Should be fixed by #26.
